include/up-cpp/communication/RpcClient.h

```diff
@@ -0,0 +1,101 @@
+// SPDX-FileCopyrightText: 2024 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache License Version 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef UP_CPP_CLIENT_RPCCLIENT_H
+#define UP_CPP_CLIENT_RPCCLIENT_H
+
+#include <up-cpp/datamodel/builder/Payload.h>
+#include <up-cpp/datamodel/builder/UMessage.h>
+#include <up-cpp/transport/UTransport.h>
+#include <up-cpp/utils/Expected.h>
+#include <uprotocol/v1/umessage.pb.h>
+#include <uprotocol/v1/ustatus.pb.h>
+
+#include <future>
+#include <memory>
+#include <string>
+#include <variant>
+
+namespace uprotocol::communication {
+
+/// @brief Interface for uEntities to invoke RPC methods.
+///
+/// Like all L2 client APIs, the RpcClient is a wrapper on top of the L1
+/// UTransport API; in this instance, it is the request-initiating half of the
+/// RPC model.
+struct RpcClient {
+	/// @brief Constructs a client connected to a given transport
+	///
+	/// @param method UUri of the method this client will invoke.
+	/// @param priority Priority of request messages when invoking methods.
+	/// @param ttl Time from the moment `invokeMethod()` is called that the
+	///            request will still be considered valid. Must be >0.
+	/// @param payload_format (Optional) If sending a payload, this sets the
+	///                       format that will be expected when invokeMethod()
+	///                       is called. Empty request payloads can only be
+	///                       sent if this was not set.
+	/// @param permission_level Permission level of this client
+	/// @param token TAP token for accessing restricted services
+	///
+	/// For guidance on the permeission_level and token parameters, see:
+	/// https://github.com/eclipse-uprotocol/up-spec/blob/main/basics/permissions.adoc
+	explicit RpcClient(std::shared_ptr<transport::UTransport> transport,
+	                   v1::UUri&& method, v1::UPriority priority,
```

Comment:
What do you think about for the constructors for the uP-L2 objects taking a configuration parameter, e.g. `RpcClientConfig`?

I see the parameter list growing a bit on this one in particular growing a bit long.

Not clear if this is considered good form or not, though.

Reply:
I have mixed feelings on it. Most likely it'll turn a call like this:
```
RpcClient foo(transport, methodUri, UPRIORITY_CS4, 100ms);
```
and turn it into this:
```
RpcClient foo({.transport=transport, .method=methodUri, .priority=UPRIORITY_CS4, .ttl=100ms});
```
I'm not sure that significantly improves readability in this case.

Reply:
Ah I see what you mean. Fair point.